src/main/java/me/karakelley/http/controllers/Application.java

```diff
@@ -3,26 +3,36 @@
 import me.karakelley.http.PublicDirectory;
 import me.karakelley.http.Request;
 import me.karakelley.http.Response;
+import me.karakelley.http.utility.HtmlGenerator;
 
 public class Application implements Controller {
   private final Router router = new Router();
+  private PublicDirectory publicDirectory = null;
 
   public Application() {
     router.route("/", new HelloWorldController());
     setupCommonRoutes();
   }
 
   public Application(PublicDirectory publicDirectory) {
-    router.route("/", new StaticFilesController());
+    this.publicDirectory = publicDirectory;
+    router.route("/", new StaticFilesController(publicDirectory, new HtmlGenerator()));
     setupCommonRoutes();
   }
 
   @Override
   public Response respond(Request request) {
+    checkRequestedPath(request);
     return router.respond(request);
   }
 
   private void setupCommonRoutes() {
     router.route("/redirectme", new RedirectController());
   }
+
+  private void checkRequestedPath(Request request) {
+    if (publicDirectory != null && publicDirectory.resourceExists(request.getPath())) {
+      router.route(request.getPath(), new StaticFilesController(publicDirectory, new HtmlGenerator()));
```

Comment:
I was not expecting to see routes dynamically added to the controller every time a request comes in.... especially in a method called `checkRequestedPath`. Why do you need to do this?

Reply:
I thought it made sense to create the routes as they come in - Initially, I created the function in `Router` to avoid hitting I/O every time thus I was creating dynamic routes as requests for valid resources came in. I realize now, after moving it to `Application` that it is no longer serving the purpose I intended. I currently see two scenarios, either add routes dynamically as they come in or check if the file exists every time no matter what.

Reply:
Ahhhh. I see. I think the avoiding IO on every request is a good thing to think about. There are probably cleaner ways to go about it, though. Can you think of a way you could use the [Proxy Pattern](https://en.wikipedia.org/wiki/Proxy_pattern) to avoid the IO?

Reply:
After reading about it, I think I understand how the pattern works but what I don't understand is how using it avoids the IO. Eventually you're still hitting IO to check if the resource exists, right?

Reply:
https://en.wikipedia.org/wiki/Memoization

Reply:
Does that spark any ideas?